STUDIO/org.openl.rules.repository.azure/src/org/openl/rules/repository/azure/AzureBlobRepository.java

```diff
@@ -0,0 +1,760 @@
+import com.azure.storage.common.implementation.Constants;
+import org.apache.commons.collections4.map.PassiveExpiringMap;
+import org.openl.rules.repository.api.*;
+import org.openl.rules.repository.common.ChangesMonitor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.yaml.snakeyaml.TypeDescription;
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.constructor.Constructor;
+import org.yaml.snakeyaml.representer.Representer;
+
+public class AzureBlobRepository implements FolderRepository {
+    private static final String UNSUPPORTED_IN_FOLDER_REPOSITORY = "Unsupported in folder repository";
+    private static final String VERSION_FILE = "versions.yaml";
+
+    private static final String MODIFICATION_FILE = "[openl]/.modification";
+    private static final String CONTENT_PREFIX = "[content]/";
+    private static final String VERSIONS_PREFIX = "[openl]/versions/";
+
+    private final Logger log = LoggerFactory.getLogger(AzureBlobRepository.class);
+
+    private String id;
+    private String name;
+    private ChangesMonitor monitor;
+    private int listenerTimerPeriod = 10;
+    private String uri;
+
+    private BlobContainerClient blobContainerClient;
+    private PassiveExpiringMap<CacheKey, AzureCommit> commitsCache;
+
+    public void setId(String id) {
+        this.id = id;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public void setMonitor(ChangesMonitor monitor) {
+        this.monitor = monitor;
+    }
+
+    public void setListenerTimerPeriod(int listenerTimerPeriod) {
+        this.listenerTimerPeriod = listenerTimerPeriod;
+    }
+
+    public void setUri(String uri) {
+        this.uri = uri;
+    }
+
+    public void initialize() {
+        blobContainerClient = new BlobContainerClientBuilder()
+                .endpoint(uri)
+                .buildClient();
+
+        commitsCache = new PassiveExpiringMap<>(10, TimeUnit.SECONDS);
+
+        monitor = new ChangesMonitor(this::getLatestRevision, listenerTimerPeriod);
+    }
+
+    @Override
+    public List<FileData> listFolders(String path) throws IOException {
+        try {
+            final String pathPrefix = VERSIONS_PREFIX + path;
+
+            List<FileData> folders = new ArrayList<>();
+
+            AzureCommit commit = findCommit(path, null);
+            if (commit != null) {
+                // Get sub-folders inside the project.
+                final int start = path.length();
+                Set<String> subFolders = new HashSet<>();
+                List<FileInfo> files = commit.getFiles();
+                if (files == null) {
+                    return folders;
+                }
+                for (FileInfo file : files) {
+                    String name = file.getPath();
+                    if (name.startsWith(path)) {
+                        String subFolder = name.substring(start, name.indexOf('/', start));

```

Comment:
Think about of using of java.nio.Path for all operation with file's paths, like, ``path.relativize(name)``. It will make all operations with paths more clear and predictive

Reply:
In general, I would agree, but in this case I think that java.nio.Path isn't good option.
Path uses current OS filesystem, so if server runs on Windows, it will expect that path separator is "\".
In Azure Blob Repository we always use "/" as a separator for a folder emulation.

Reply:
I understand your point, but it's possible to use the following function https://github.com/openl-tablets/openl-tablets/blob/master/DEV/org.openl.commons/src/org/openl/util/formatters/FileNameFormatter.java#L12 to fix path to UNIX style before assigning into DTO